[jarrellc1] iP

jarrellc1 opened the pull request:
**Jar**
> "Faster" – Noone

Jar frees your mind from having to remember things you need to do. It's:

text-based
easy to learn
~~FAST~~ SUPER FAST to use
All you need to do is:

- Download it from [here](https://github.com/jarrellc1/ip/releases).
- Double-click it.
- Add your tasks.
- Let it manage your tasks for you :smile:.

Features:

1. Todo, Deadline and event tasks with dates
2. Find, delete and list

##Yet to do:
- [x] Further Commands

```java
public class Main {
    public static void main(String[] args) {
        Application.launch(MainApp.class, args);
    }
}

Commit:
Add Gradle support

Eclipse-Dominator committed:
Bump gradle and lib version

aureliony committed:
build.gradle: Prevent generating a second JAR file

In build.gradle, the dependencies on distZip and/or distTar causes
the shadowJar task to generate a second JAR file for which the
mainClass.set("seedu.duke.Duke") does not take effect.
Hence, this additional JAR file cannot be run.
For this product, there is no need to generate a second JAR file
to begin with.

Let's remove this dependency from the build.gradle to prevent the
shadowJar task from generating the extra JAR file.

Commit:
Level-0

Commit:
Level-1

Commit:
Level-2

Commit:
Level-3

Commit:
Level-4

Commit:
A-TextUiTesting

-added automated testing sh file.

Commit:
Level-5

-added exception handling

Commit:
Parser class: Add isDelete() method and delete logic in handleCommand method

This is to check if the command contains a delete, which will initiate the logic for delete tasks.

Ui class: Add delete Ui

This is to display to user delete messages when a successful or failed delete is initiated.

TaskList class: Add delete method

Supporting methods using Java Collections for deleting task in TaskList.

Commit:
Add folder for enums
Add CommandName.java class in folder enums

Using enums for command names, we can implement a cleaner way to store and identify the different
command names. We will implement the logic by refactoring the Parser.java class.

Parser class: Refactored code, deleted boolean methods such as isDelete(), and used switch statements

Code simplicity and avoid duplication of code by using switch statements.
This allows for easier scalability and maintenance of code logic.

ToDo, DeadLine, and Event switch statements have fallthrough to prevent code duplication.

Commit:
Add Storage class

Storage class will contain the logic for writing the file into the user device and will
also load existing user data.

Add toFileFormat() to task, event, deadLine and toDo classes to allow code to write
in a specific format to save into the device.

We will save the file in the Jar class once user inputs "bye".

Commit:
Add dates and times functionality to the chatbot
Add java.time.LocalDate to DeadLine
Rename Deadline class to DeadLine
Refactor Storage class to read new file format for DeadLine command

Commit:
Refactor code for Parser and Ui

Commit:
Refactor code for parser

Commit:
Add packages for tasks, parser, exceptions, storage, and ui
Refactor call the affected class code to accommodate packages

This is to allow us to organize the classes into suitable Java packages.

Commit:
Merge commit '81a9c537afe8492675eb253968939364d65db091'

Commit:
Add comments to JarException

Testing commit.

Ignore this commit.

Commit:
Add ParserTest.java and StorageTest.java in Test folder
Refactor Storage.java, as it was throwing the wrong exception

ParserTest will test the handleCommend method in Parser class and StorageTest will test for
proper loading of files and corrupted files.

Commit:
Update build.gradle by changing main class set to Jar

Commit:
Add javadocs for all classes for Jar bot

Documented class-level, constructor-level, and method-level details.
Improved code readability and maintainability by providing clear explanations of class purposes, methods, and their usage.

Commit:
Refactor code to follow java coding standard
Remove unused Ui statements

Commit:
Add find enum to CommandName
Add find functionality to TaskList
Add find logic to handleCommand in Parser
Add showTaskList to Ui

This is to allow the user to input find and a keyword to search list for tasks containing
the substring of the task wanted. ShowTaskList in Ui is an overloaded method to display
the array of found tasks for the user.

Commit:
Merge branch 'branch-A-JavaDoc'

Commit:
Merge branch 'branch-Level-9'

# Conflicts:
#	src/main/java/ui/Ui.java

Commit:
Refactor Ui code

Commit:
Refactor code for Parser, Storage, DeadLine, TaskList, ToDo and Ui

This is to follow the Java coding standard, mainly refactored switch statements
and ensure that the right margin of 120 chars is not exceeded.

Commit:
Add JavaFX GUI for Jarbot
Add folder gui containing classes DialogBox, Main, MainWindow
and Launcher for JavaFX
Add resources folder containing sub-packages of images and view,
containing DialogBox.fxml and MainWindow.fxml
Refactor Jar.java to contain its own package for code neatness
Refactor Ui class to return string for all messages methods
Refactor CommandName to return string of the messages for
each command

These changes are made to facilitate the GUI logic, and CLI logic
has been removed.

e1121208 commented on src/main/java/enums/CommandName.java:
```diff
@@ -0,0 +1,17 @@
+package enums;
+
+/**
+ * Represents the different types of commands that can be issued to the jar.Jar Bot.
+ */
+public enum CommandName {
+    TODO,
+    DEADLINE,
+    EVENT,
+    MARK,
+    UNMARK,
+    DELETE,
+    LIST,
+    EXIT,
+    FIND,
+    UNKNOWN
+}
```
Great use of enums!

e1121208 reviewed (commented):
Really not much I can say than well done! Variable names generally follow standard conventions, good job!

blackpanther9229 commented on src/main/java/parser/Parser.java:
```diff
@@ -0,0 +1,195 @@
+     * @return The corresponding Task object.
+     * @throws JarException If the command is unknown or incorrectly formatted.
+     */
+    public Task parseTask(String command) throws JarException {
```
I think method name should be `taskParser`

blackpanther9229 commented on src/main/java/parser/Parser.java:
```diff
@@ -0,0 +1,195 @@
+     * @param command The command string to be interpreted.
+     * @return The corresponding CommandName enum value.
+     */
+    public CommandName parseCommandType(String command) {
```
I think method name should be `commandTypeParser`

blackpanther9229 commented on src/main/java/parser/Parser.java:
```diff
@@ -0,0 +1,195 @@
+     * @return A string containing the response to the command.
+     * @throws JarException If the command is unknown or if an error occurs during command handling.
+     */
+    public String handleCommand(String command, TaskList taskList, Ui ui) throws JarException {
```
To be more clear, I suggest your method name to be something like `getResponseFromCommand`

blackpanther9229 commented on src/main/java/tasks/DeadLine.java:
```diff
@@ -0,0 +1,45 @@
+ * Represents a task with a deadline.
+ * A DeadLine task has a description and a specific date and time by which it must be completed.
+ */
+public class DeadLine extends Task {
```
I prefer `Deadline` because it is a single word, but it is up to you

blackpanther9229 commented on src/main/java/tasks/DeadLine.java:
```diff
@@ -0,0 +1,45 @@
+     * @return A string formatted for saving to a file.
+     */
+    @Override
+    public String toFileFormat() {
```
Clear and starightfoward method name. Good job!

blackpanther9229 commented on src/main/java/tasks/TaskList.java:
```diff
@@ -0,0 +1,156 @@
+     *
+     * @return The list of tasks.
+     */
+    public ArrayList<Task> getTasks() {
```
To be more clear, I suggest the method name to be `getListOfTasks`

blackpanther9229 commented on src/main/java/tasks/TaskList.java:
```diff
@@ -0,0 +1,156 @@
+        if (keyword == null || keyword.trim().isEmpty()) {
+            throw new JarException("Task cannot be empty!");
+        } else {
+            ArrayList<Task> sameTasks = new ArrayList<>();
```
`sameTasks` is too vague. I suggest `filteredTasks`

jarrellc1 replied:
Thanks for the suggestion.

blackpanther9229 commented on src/main/java/ui/Ui.java:
```diff
@@ -0,0 +1,149 @@
+     * @param task The task that was marked as completed.
+     * @return The task marked message as a string.
+     */
+    public String showTaskMarked(Task task) {
```
This method name should be more proper: `markCommandResponse` and so on for the rest of the methods

jarrellc1 replied:
Thanks for the comment.

YangQF2002 commented on src/main/java/parser/Parser.java:
```diff
@@ -0,0 +1,195 @@
+            taskList.deleteTask(deleteNumber);
+            response = ui.showDeleteTask(deleteTask) + "\n" + ui.showTaskCount(taskList.getTaskCount());
+            break;
+        case TODO:
+            // Fallthrough
+        case DEADLINE:
+            // Fallthrough
```
Nice fallthrough comment here for non-break cases 👍

YangQF2002 commented on src/main/java/tasks/DeadLine.java:
```diff
@@ -0,0 +1,45 @@
+     *
+     * @return A string formatted for saving to a file.
+     */
+    @Override
+    public String toFileFormat() {
+        DateTimeFormatter formatter = DateTimeFormatter.ofPattern("d/M/yyyy HHmm");
+        return "D | " + super.toFileFormat() + " | " + by.format(formatter);
+    }
```
Nice use of OOP (inheritance) to convert task object to string format 👍 
Maybe consider a line break before the second + operator for slightly better code readability

YangQF2002 commented on src/main/java/tasks/Event.java:
```diff
@@ -0,0 +1,42 @@
+     *
+     * @return A string representation of the event.
+     */
+    @Override
+    public String toString() {
+        return "[E]" + super.toString() + " (from: " + from + " to: " + to + ")";
+    }
+
+    /**
+     * Returns a string format of the Event task for saving to a file.
+     *
+     * @return A string formatted for saving to a file.
+     */
+    @Override
+    public String toFileFormat() {
+        return "E | " + super.toFileFormat() + " | " + from + " | " + to;
+    }
+}
```
Same suggestion here regarding line breaks

YangQF2002 commented on src/main/resources/images/DaDuke.png:
Stunning images 💯

YangQF2002 commented on src/test/java/storage/StorageTest.java:
```diff
@@ -0,0 +1,81 @@
+        assertEquals("[D][ ] Submit assignment (by: Aug 24 2023, 23:59)", loadedTasks.get(1).toString());
+    }
+
+    @Test
+    public void load_corruptedFile_throwsJarException() throws IOException {
+        try (FileWriter writer = new FileWriter(TEST_FILE_PATH)) {
+            writer.write("T | 1 | Read book\n"); // valid line
+            writer.write("X | 1 | Submit assignment | 2/12/2023 2359\n"); // Invalid task type, X
+            writer.write("sjfsjfhjsnl"); // Corrupted line
+        }
+
+        JarException exception = assertThrows(JarException.class, () -> {
+            storage.load();
+        });
+
+        assertEquals("Data file corrupted. Invalid task type.", exception.getMessage());
+    }
+
+}
```
Nice test here to test for corrupted save file

YangQF2002 reviewed (commented):
Overall, LGTM, nice OOP structure. Just consider using line breaks to improve overall code readability.

Commit:
Add assertions and improve error handling in core Jar Bot classes

Added assertions to key parts of the Storage, Parser, and Jar classes to
ensure important assumptions hold true during runtime.

There are several places where assertions have been added. For storage, we will
verify the file path and task list before saving or loading. In Parser, we will ensure non-null
input commands and task numbers. In Jar, we will ensure that getResponse(String input),
input is not null or empty. These assertions make the code more robust can help
identify any bugs, and do not affect the code functionality.

Commit:
Refactor Parser class parseTask(String command) method

We will avoid long methods by extracting the logic of parseTask
to separate methods such as parseToDoTask, parseDeadlineTask
and parseEventTask.

Commit:
Refactor Parser class to maintain readability

Let's extract the logic of handleCommand method to different sub methods
so that handleCommand method will not be too long.

jarrellc1 committed:
Merge pull request #1 from jarrellc1/branch-A-Assertions

Add assertions and improve error handling in core Jar Bot classes

Commit:
Merge branch 'master' into branch-A-CodeQuality

Commit:
Refactor Storage class load() method to adhere to code quality

Let's extract the logic inside load method to ensure that the method length
does not exceed 30 lines, and to ensure readibility.

jarrellc1 committed:
Merge pull request #2 from jarrellc1/branch-A-CodeQuality

Refactor Storage class load() method to adhere to code quality

Commit:
Add find duplicate logic to Jar bot

Let's create a method called findDuplicate in TaskList, which will check if the task entered
exists in the list of tasks. If so, we will return a message to user indicating that they cannot add
the task.

Let's also add some getter and setter methods in the task classes to extract information from the tasks
while adhering to OOP conventions.

jarrellc1 committed:
Merge pull request #3 from jarrellc1/branch-C-DetectDuplicates

Add find duplicate logic to Jar bot

Commit:
Add JavaDocs for methods in these classes that handle the new functionality
of finding duplicate tasks

Commit:
Refactor TaskList class for improved readability and maintainability

Remove taskList from findDuplicate method.

Let's remove taskList from findDuplicate method as it is redundant as findDuplicate
is already calling from taskList and does not need another copy of the list of tasks

Commit:
Refactor Parser and TaskList code

Let's change the code in the classes to adhere to java coding standard
by including line break to ensure line does not cross 120 char limit.

Commit:
Refactor Storage class variable length to improve readability

Let's add line break for DATE_TIME_FORMATTER so that it does not exceed
character limit.

Commit:
Refactor task classes to maintain readability

Let's create a constant variable called TASK_TYPE which will return a
string describing the task type. The getter method will return TASK_TYPE.

Commit:
Add javadocs for gui classes

Let's add javadocs for these classes to explain the purpose of the DialogBox GUI component.

Commit:
Refactor task parsing logic in Storage class to improve readability

Let's extract the logic of task parsing out into shorter methods. ParseTaskStatus
is to check the task status, while createTaskFromType returns a new task according to
the type needed.

Commit:
Add User Guide to README.md and Ui.png

Let's add user guide to readme which covers all the basic functionalities, their command,
sample input and output. This is to allow user to have a guide on how to run the bot and
the different commands that they can use in this bot. Ui.png is to show how the interface
looks like for more detail.

Commit:
Add User Guide to README.md and Ui.png

Let's add user guide to readme which covers all the basic functionalities, their command,
sample input and output. This is to allow user to have a guide on how to run the bot and
the different commands that they can use in this bot. Ui.png is to show how the interface
looks like for more detail.

Commit:
Refactor Parser and Ui class to maintain readability

Let's refactor these classes to prevent deeply nested functions for
code readability.

Commit:
Refactor README.md to make it more readable

Let's change the heading so that it looks more user friendly.

Commit:
Refactor ReadMe.md to ensure readability

Let's continue refactoring ReadMe to ensure that it comply with UserGuide.

Commit:
Add checkstyle xml files

Let's add checkstyle xml files to help with style checking.

Commit:
Refactor all files to adhere to checkstyle and coding style

Let's refactor all the classes that have checkstyle violations so that it
becomes more readable.
